Report generation fails with null underlying for option orders

#### Expected Behavior
When running report generator using downloaded JSON results from a QuantConnect platform run backtest with options traded, report should be generated

#### Actual Behavior
Report generation errors with a null reference as it seems the underlying of the option is not being deserialized. Because a loaded option symbol has a null underlying reference it fails at SecurityService.cs line 131:

`Unhandled Exception:
System.NullReferenceException: Object reference not set to an instance of an object
  at QuantConnect.Securities.SecurityService.CreateSecurity (QuantConnect.Symbol symbol, System.Collections.Generic.List`1[T] subscriptionDataConfigList, System.Decimal leverage, System.Boolean addToSymbolCache) [0x00272] in /home/ari/Development/projects/Lean/Common/Securities/SecurityService.cs:131 
  at QuantConnect.Securities.SecurityManager.CreateSecurity (QuantConnect.Symbol symbol, System.Collections.Generic.List`1[T] subscriptionDataConfigList, System.Decimal leverage, System.Boolean addToSymbolCache) [0x00001] in /home/ari/Development/projects/Lean/Common/Securities/SecurityManager.cs:302 
  at QuantConnect.Report.PortfolioLooperAlgorithm.FromOrders (System.Collections.Generic.IEnumerable`1[T] orders) [0x0006c] in /home/ari/Development/projects/Lean/Report/PortfolioLooper/PortfolioLooperAlgorithm.cs:45 
  at (wrapper remoting-invoke-with-check) QuantConnect.Report.PortfolioLooperAlgorithm.FromOrders(System.Collections.Generic.IEnumerable`1<QuantConnect.Orders.Order>)
  at QuantConnect.Report.PortfolioLooper..ctor (System.Double startingCash, System.Collections.Generic.List`1[T] orders, QuantConnect.Resolution resolution) [0x001ca] in /home/ari/Development/projects/Lean/Report/PortfolioLooper/PortfolioLooper.cs:122 
  at QuantConnect.Report.PortfolioLooper+<FromOrders>d__13.MoveNext () [0x00283] in /home/ari/Development/projects/Lean/Report/PortfolioLooper/PortfolioLooper.cs:270 
  at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in /builddir/build/BUILD/mono-6.12.0.90/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs:1108 
  at System.Collections.Generic.List`1[T]..ctor (System.Collections.Generic.IEnumerable`1[T] collection) [0x00062] in /builddir/build/BUILD/mono-6.12.0.90/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs:87 
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0000e] in /builddir/build/BUILD/mono-6.12.0.90/external/corefx/src/System.Linq/src/System/Linq/ToCollection.cs:30 
  at QuantConnect.Report.Report..ctor (System.String name, System.String description, System.String version, QuantConnect.Packets.BacktestResult backtest, QuantConnect.Packets.LiveResult live) [0x00080] in /home/ari/Development/projects/Lean/Report/Report.cs:50 
  at QuantConnect.Report.Program.Main (System.String[] args) [0x0012c] in /home/ari/Development/projects/Lean/Report/Program.cs:73 
[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at QuantConnect.Securities.SecurityService.CreateSecurity (QuantConnect.Symbol symbol, System.Collections.Generic.List`1[T] subscriptionDataConfigList, System.Decimal leverage, System.Boolean addToSymbolCache) [0x00272] in /home/ari/Development/projects/Lean/Common/Securities/SecurityService.cs:131 
  at QuantConnect.Securities.SecurityManager.CreateSecurity (QuantConnect.Symbol symbol, System.Collections.Generic.List`1[T] subscriptionDataConfigList, System.Decimal leverage, System.Boolean addToSymbolCache) [0x00001] in /home/ari/Development/projects/Lean/Common/Securities/SecurityManager.cs:302 
  at QuantConnect.Report.PortfolioLooperAlgorithm.FromOrders (System.Collections.Generic.IEnumerable`1[T] orders) [0x0006c] in /home/ari/Development/projects/Lean/Report/PortfolioLooper/PortfolioLooperAlgorithm.cs:45 
  at (wrapper remoting-invoke-with-check) QuantConnect.Report.PortfolioLooperAlgorithm.FromOrders(System.Collections.Generic.IEnumerable`1<QuantConnect.Orders.Order>)
  at QuantConnect.Report.PortfolioLooper..ctor (System.Double startingCash, System.Collections.Generic.List`1[T] orders, QuantConnect.Resolution resolution) [0x001ca] in /home/ari/Development/projects/Lean/Report/PortfolioLooper/PortfolioLooper.cs:122 
  at QuantConnect.Report.PortfolioLooper+<FromOrders>d__13.MoveNext () [0x00283] in /home/ari/Development/projects/Lean/Report/PortfolioLooper/PortfolioLooper.cs:270 
  at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in /builddir/build/BUILD/mono-6.12.0.90/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs:1108 
  at System.Collections.Generic.List`1[T]..ctor (System.Collections.Generic.IEnumerable`1[T] collection) [0x00062] in /builddir/build/BUILD/mono-6.12.0.90/external/corefx/src/Common/src/CoreLib/System/Collections/Generic/List.cs:87 
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0000e] in /builddir/build/BUILD/mono-6.12.0.90/external/corefx/src/System.Linq/src/System/Linq/ToCollection.cs:30 
  at QuantConnect.Report.Report..ctor (System.String name, System.String description, System.String version, QuantConnect.Packets.BacktestResult backtest, QuantConnect.Packets.LiveResult live) [0x00080] in /home/ari/Development/projects/Lean/Report/Report.cs:50 
  at QuantConnect.Report.Program.Main (System.String[] args) [0x0012c] in /home/ari/Development/projects/Lean/Report/Program.cs:73 `

An example piece of relevant JSON downloaded from QuantConnect looks like:
`"Symbol": {
          "Value": "WAG   101016P00025000",
          "ID": "WAG 2YPXFKVWULF1I|WAG R735QTJ8XC9X",
          "Permtick": "WAG   101016P00025000",
          "Underlying": {
            "Value": "WAG",
            "ID": "WAG R735QTJ8XC9X",
            "Permtick": "WAG"
          }
        }`

#### Potential Solution

#### Reproducing the Problem
Run a backtest that trades options (in my case I only traded options, and not the underlying, so this may be related) on QuantConnect, then download the JSON of the total results. Run the report generator on that file.

#### System Information
Linux/mono

#### Checklist
- [X] I have completely filled out this template
- [X] I have confirmed that this issue exists on the current `master` branch
- [X] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
<!--- Required for Bugs, feature request can delete the line below. -->
- [X] I have provided detailed steps to reproduce the issue

--- a/Common/Symbol.cs
+++ b/Common/Symbol.cs
@@ -307,6 +307,10 @@ namespace QuantConnect
                 throw new ArgumentNullException(nameof(value));
             }
             ID = sid;
+            if (ID.HasUnderlying)
+            {
+                Underlying = new Symbol(ID.Underlying, ID.Underlying.Symbol);
+            }
 
             Value = value.LazyToUpper();
         }

--- a/Report/PortfolioLooper/PortfolioLooper.cs
+++ b/Report/PortfolioLooper/PortfolioLooper.cs
@@ -128,6 +128,9 @@ namespace QuantConnect.Report
             // More initialization, this time with Algorithm and other misc. classes
             _resultHandler.Initialize(job, new Messaging.Messaging(), new Api.Api(), transactions);
             _resultHandler.SetAlgorithm(Algorithm, Algorithm.Portfolio.TotalPortfolioValue);
+
+            Algorithm.Transactions.SetOrderProcessor(transactions);
+
             transactions.Initialize(Algorithm, new BacktestingBrokerage(Algorithm), _resultHandler);
             feed.Initialize(Algorithm, job, _resultHandler, null, null, null, _dataManager, null, null);
 

--- a/Tests/Common/SymbolTests.cs
+++ b/Tests/Common/SymbolTests.cs
@@ -587,6 +587,23 @@ namespace QuantConnect.Tests.Common
             Assert.IsTrue(nonCanonicalFutureOption.Value.StartsWith(expectedFutureOptionTicker));
         }
 
+        [Test]
+        public void SymbolWithSidContainingUnderlyingCreatedWithoutNullUnderlying()
+        {
+            var future = Symbol.CreateFuture("ES", Market.CME, new DateTime(2020, 6, 19));
+            var optionSid = SecurityIdentifier.GenerateOption(
+                future.ID.Date,
+                future.ID,
+                future.ID.Market,
+                3500m,
+                OptionRight.Call,
+                OptionStyle.American);
+
+            var option = new Symbol(optionSid, "ES");
+            Assert.IsNotNull(option.Underlying);
+            Assert.AreEqual(future, option.Underlying);
+        }
+
         class OldSymbol
         {
             public string Value { get; set; }

--- a/Tests/Report/PortfolioLooperTests.cs
+++ b/Tests/Report/PortfolioLooperTests.cs
@@ -113,5 +113,70 @@ namespace QuantConnect.Tests.Report
             Assert.AreEqual(80000, pointInTimePortfolio[1].TotalPortfolioValue);
             Assert.AreEqual(80000, pointInTimePortfolio[2].TotalPortfolioValue);
         }
+
+        [Test]
+        public void OptionOrderDoesNotThrow()
+        {
+            var equityPoints = new SortedList<DateTime, double>
+            {
+                { new DateTime(2019, 1, 3, 5, 0, 5), 100000 },
+                { new DateTime(2019, 1, 4, 5, 0, 5), 90000 },
+            };
+
+            var series = new Series<DateTime, double>(equityPoints);
+            var equity = Symbol.Create("SPY", SecurityType.Equity, Market.USA);
+            var optionSid = SecurityIdentifier.GenerateOption(
+                equity.ID.Date,
+                equity.ID,
+                equity.ID.Market,
+                200m,
+                OptionRight.Call,
+                OptionStyle.American);
+            var option = new Symbol(optionSid, optionSid.Symbol);
+
+            var entryOrder = Order.CreateOrder(new SubmitOrderRequest(
+                OrderType.Market,
+                SecurityType.Option,
+                option,
+                1,
+                0m,
+                0m,
+                new DateTime(2019, 1, 3, 5, 0, 5),
+                string.Empty
+            ));
+            var exitOrder = Order.CreateOrder(new SubmitOrderRequest(
+                OrderType.Market,
+                SecurityType.Option,
+                option,
+                -1,
+                0m,
+                0m,
+                new DateTime(2019, 1, 4, 5, 0, 5),
+                string.Empty
+            ));
+
+            entryOrder.LastFillTime = new DateTime(2019, 1, 3, 5, 0, 5);
+            exitOrder.LastFillTime = new DateTime(2019, 1, 4, 5, 0, 5);
+
+            entryOrder.GetType().GetProperty("Id").SetValue(entryOrder, 1);
+            entryOrder.GetType().GetProperty("Price").SetValue(entryOrder, 100000m);
+
+            Order marketOnFillOrder = null;
+            exitOrder.GetType().GetProperty("Id").SetValue(exitOrder, 2);
+            exitOrder.GetType().GetProperty("Price").SetValue(exitOrder, 80000m);
+            exitOrder.GetType().GetProperty("Status").SetValue(exitOrder, OrderStatus.Filled);
+
+            var orders = new[] { entryOrder, marketOnFillOrder, exitOrder }.Where(x => x != null);
+
+            var looper = PortfolioLooper.FromOrders(series, orders);
+            Assert.DoesNotThrow(() =>
+            {
+                foreach (var pointInTimePortfolio in looper)
+                {
+                    Assert.AreEqual(option, pointInTimePortfolio.Order.Symbol);
+                    Assert.AreEqual(option.Underlying, pointInTimePortfolio.Order.Symbol.Underlying);
+                }
+            });
+        }
     }
 }

